chore: Upgrade to React Native 0.76

thehale opened the pull request:
This has been an intensely difficult upgrade.

I attempted to use the upgrade guide, but ran into problem after
problem. The approach that ultimately proved successful was to
scaffold a brand new app, copy in my old code, and make
adjustments as needed to fix bugs.

Each commit represents a logical progression through the upgrade.
They are an idealized path someone else could follow, learning from
the many mistakes/frustrations I faced along the way.

thehale committed:
chore: Upgrade `react-navigation` to the New Architecture

Includes transitive upgrades to `react-native-reanimated` and all
the navigators used in the app. Removes a few unused packages.

thehale committed:
chore: Upgrade `react-native-fs` to the New Architecture

The mainline `react-native-fs` hasn't had a commit for almost
18 months and hasn't had a new release for over 2 years.

So, I switched to the most stable looking fork which has been
updated much more regularly and has a decent sized user base.

thehale committed:
chore: Upgrade `react-native-ble-manager` to the New Architecture

This commit looks deceptively simple. It was one of the hardest
parts of the upgrade.

First, `LocationStatus.ts` was part of a custom native module I
had written to support the Android versions which require
location permissions for Bluetooth scans. Since that was on the
Old Architecture, I had to learn how to write a New Architecture
native module. That ended up being easiest to do by publishing
a separate package on npm.

Additionally, the maintainers of `react-native-ble-manager` are
still working on full support for the New Architecture. Timing was
very much in my favor since they started publishing viable release
candidates mere days before I started this upgrade project, and
the full release was published only ~20 hours ago.

thehale committed:
chore: Upgrade `realm` to the New Architecture

MongoDB, the company that acquired Realm several years ago,
recently announced that they are deprecating Realm :/

So, I'll need to start building an alternate mass storage solution.
Fortunately, it looks like there's a new kid on the sqlite block,
`react-native-nitro-sqlite`, so I may be able to switch to the
database I had wanted to use in the first place.

BUT... that migration doesn't need to happen today. Realm still
works for now, so we'll keep it and upgrade it until we are able
to move to a more stable datastore.

thehale committed:
chore: Upgrade remaining packages to latest supported versions

Getting everything else up to date as much as possible.

Some additional notes
===============
`react-native-charts-wrapper` - Will need to replace this soon
too. It hasn't moved to the new architecture yet and probably
won't any time soon (no recent commits/updates). The best
alternative I'm seeing right now is `react-native-echarts` with the
`react-native-skia` renderer.

`react-native-paper` - The lead maintainer of this component
library has moved to other priorites at Callstack. Updates here,
while not stalled, have dramatically slowed down. Replacing the
entire component library will be a big lift, so I'm not eager to do
this yet, but I may have to somewhat soon.

When it comes time to really push on both of these big library
migrations, I want to put a big priority on choosing tools that
support compatibility with `react-native-windows`,
`react-native-macos`, and ideally `react-native-web`. Some of the
choices I made during this 0.76 upgrade move the app closer to
supporting a Desktop build, and I want to continue that arc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SpeedcuberOSS/speedcuber-timer?pullRequest=130) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SpeedcuberOSS/speedcuber-timer?pullRequest=130) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SpeedcuberOSS/speedcuber-timer?pullRequest=130) it.</sub>

thehale committed:
chore: Update lock files

It wasn't worth the effort to try and track how these lock files
would have changed through the "idealized" upgrade process.
So, instead, I'm just committing their final result here at the end.

thehale committed:
chore: Upgrade to React Native 0.76

This has been an intensely difficult upgrade.

I attempted to use the upgrade guide, but ran into problem after
problem. The approach that ultimately proved successful was to
scaffold a brand new app, copy in my old code, and make
adjustments as needed to fix bugs.

This commit contains the core-only changes needed for the 0.76
upgrade. Subsequent commits will contain the third-party changes
required for full functionality on 0.76

thehale committed:
chore: Upgrade `react-native-vector-icons` to New Architecture

The best part is that this upgrade **simplifies** my code and processes.

I no longer have to do the awkward manual linking steps in
Android and iOS. In fact, because `react-native-vector-icons` now
installs all fonts in `assets/fonts`, it also handles my custom fonts!

This commit also removes all the unused icon packs. My hope is
that will contribute to a smaller bundle size.

The one downside is that I have to add a bunch of mocks for my tests
to pass again. Hopefully that will be fixed more cleanly in the future.
 - See: https://github.com/oblador/react-native-vector-icons/issues/1681#issuecomment-2585560725

thehale committed:
chore: Update component snapshots

Underlying library components have updated tree structures.

I've manually validated the appearance of all these components.
Next will be additional reviews of the functionality.

thehale committed:
chore: Upgrade `react-native-webview`

Switches to the version by @dr.pogodin since, currently,
it has more patches/bug fixes implemented.

thehale committed:
fix: Point `TwistyPlayer` WebView to pinned `cubing.js`

The cubing.js update from 0.46.2 to 0.47.0 broke something in
iOS's ability to render twisty puzzles. Since the official CDN only
hosts the latest version of cubing.js, I had to build my own
website to load into the WebView instead.

That website is found at https://github.com/SpeedcuberOSS/twisty-player

thehale committed:
chore: Upgrade to React Native `0.76.9`

The key motivator for this upgrade was an issue with the newly released Xcode 16.3 which broke existing iOS builds.

See: https://github.com/facebook/react-native/issues/50411
